momaland/envs/connect4/connect4.py

```diff
@@ -0,0 +1,407 @@
+or not; the action encoding is described in the Action Space section below.
+The `action_mask` will be all zeros for any agent except the one whose turn it is. Taking an illegal action ends the
+game with a reward of -1 for the illegally moving agent and a reward of 0 for all other agents. #TODO this isn't happening anymore because of missing TerminateIllegalWrapper
+
+
+### Action Space
+
+The action space is the set of integers from 0 to board_width (exclusive), where the number represents which column
+a token should be dropped in.
+
+
+### Rewards
+
+Dimension 0: If an agent successfully connects four of their tokens, they will be rewarded 1 point. At the same time,
+the opponent agent will be awarded -1 point. If the game ends in a draw, both players are rewarded 0.
+Dimension 1: If an agent wins, they get a reward of 1-(move_count/board_size) to incentivize faster wins. The losing opponent gets the negated reward. In case of a draw, both agents get 0.
+Dimension 2 to board_width+1 (default 8): (optional) If at game end, an agent has more tokens than their opponent in
+column X, they will be rewarded 1 point in reward dimension 2+X. The opponent agent will be rewarded -1 point. If the
+column has an equal number of tokens from both players, both players are rewarded 0.
+
+
+### Version History
+
+"""
+from __future__ import annotations
+
+import os
+from typing_extensions import override
+
+import gymnasium
+import numpy as np
+import pygame
+from gymnasium import spaces
+from gymnasium.utils import EzPickle
+from pettingzoo.utils import agent_selector, wrappers
+
+from momaland.utils.env import MOAECEnv
+
+
+def get_image(path):
+    """Load an image into a pygame surface."""
+    from os import path as os_path
+
+    cwd = os_path.dirname(__file__)
+    image = pygame.image.load(cwd + "\\" + path)
+    sfc = pygame.Surface(image.get_size(), flags=pygame.SRCALPHA)
+    sfc.blit(image, (0, 0))
+    return sfc
+
+
+def env(**kwargs):
+    """Returns the wrapped MOConnect4 environment in `AEC` format.
+
+    Args:
+        **kwargs: keyword args to forward to the raw_env function
+
+    Returns:
+        A fully wrapped AEC env
+    """
+    env = raw_env(**kwargs)
+    # This wrapper terminates the game with the current player losing in case of illegal values.
+    # env = wrappers.TerminateIllegalWrapper(env, illegal_reward=-1)
+    # Asserts if the action given to step is outside of the action space.
+    env = wrappers.AssertOutOfBoundsWrapper(env)
+    # Checks if function calls or attribute access are in a disallowed order.
+    # env = wrappers.OrderEnforcingWrapper(env)
+    return env
+
+
+def raw_env(**kwargs):
+    """Returns the MOConnect4 environment in `AEC` format.
+
+    Args:
+        **kwargs: keyword args to forward to create the `MOConnect4` environment.
+
+    Returns:
+        A raw env.
+    """
+    return MOConnect4(**kwargs)
+
+
+class MOConnect4(MOAECEnv, EzPickle):

```

Comment:
Don't you think there is a way to extend PZ's Connect4 and just override where needed instead?

Reply:
I think so too @wilrop . I might be able to inherit one or two basic methods, but not sure that's worth it.